components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/persistence/cache/CacheBackedPolicyDAO.java

```diff
@@ -0,0 +1,178 @@
+            }
+            return policies;
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Cache miss in PapPolicyListCache for policies for tenant: %s", tenantId));
+        }
+        policies = super.getAllPAPPolicies(tenantId);
+        papPolicyListCache.addToCache(tenantId, (ArrayList<PolicyDTO>) policies, tenantId);
+        return policies;
+    }
+
+    @Override
+    public void deletePAPPolicy(String policyId, int tenantId) throws EntitlementException {
+
+        super.deletePAPPolicy(policyId, tenantId);
+        papPolicyCache.clearCacheEntry(policyId, tenantId);
+        papPolicyListCache.clearCacheEntry(tenantId, tenantId);
+    }
+
+    @Override
+    public PolicyStoreDTO getPDPPolicy(String policyId, int tenantId) {
+
+        PolicyStoreDTO policy = pdpPolicyCache.getValueFromCache(policyId, tenantId);
+        if (policy != null) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(String.format("Cache hit in PdpPolicyCache for policy: %s for tenant: %s",
+                        policyId, tenantId));
+            }
+            return policy;
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Cache miss in PdpPolicyCache for policy: %s for tenant: %s", policyId, tenantId));
+        }
+        policy = super.getPDPPolicy(policyId, tenantId);
+        pdpPolicyCache.addToCache(policyId, policy, tenantId);
+        pdpPolicyListCache.clearCacheEntry(tenantId, tenantId);
+        return policy;
+    }
+
+    @Override
+    public PolicyStoreDTO[] getAllPDPPolicies(int tenantId) throws EntitlementException {
+
+        PolicyStoreDTO[] policies = pdpPolicyListCache.getValueFromCache(tenantId, tenantId);
+        if (policies != null) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(String.format("Cache hit in PdpPolicyListCache for policies for tenant: %s", tenantId));
+            }
+            return policies;
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Cache miss in PdpPolicyListCache for policies for tenant: %s", tenantId));
+        }
+        policies = super.getAllPDPPolicies(tenantId);
+        pdpPolicyListCache.addToCache(tenantId, policies, tenantId);
+        return policies;
+    }
+
+    @Override
+    public void insertOrUpdatePolicy(PolicyStoreDTO policy, int tenantId) throws EntitlementException {
+
+        super.insertOrUpdatePolicy(policy, tenantId);
+        pdpPolicyCache.addToCache(policy.getPolicyId(), policy, tenantId);
+        pdpPolicyListCache.clearCacheEntry(tenantId, tenantId);
+    }
+
+    @Override
+    public void updateActiveStatusAndOrder(PolicyStoreDTO policy, int tenantId) throws EntitlementException {
+
+        super.updateActiveStatusAndOrder(policy, tenantId);
+        pdpPolicyCache.addToCache(policy.getPolicyId(), policy, tenantId);
```

Comment:
Do we have fully filled `PolicyStoreDTO` here?
If not we cant do a `addToCache()` with a half filled object.. Checking this because to update active status & order it doesn't sound like full `PolicyStoreDTO` needs to be filled.
(unless `PolicyStoreDTO` only contains status and order, which again contradict with having another method named `insertOrUpdatePolicy()`)

Reply:
yes, seems it is a half-filled object, will clear the cache entry from pdpPolicyCache at this point

Reply:
Ack